Porting ASP.NET eBook REVIEW DRAFT (#22167)
dotnet/docs: docs/architecture/porting-existing-aspnet-apps/example-migration-eshop.md

The initial version of the project is shown in Figure 4-1. It's a fairly standard ASP.NET MVC 5 app.

![Figure 4-1](media/Figure4-1.png)

**Figure 4-1. The *eShopModernizing* MVC sample project structure.**

## Run *ApiPort* to identify problematic APIs

The first step in preparing to migrate is to run the *ApiPort* tool. The tool identifies how many .NET Framework APIs the app calls and how many of these have .NET Standard or .NET Core equivalents. Focus primarily on your own app's logic, not third-party dependencies, and pay attention to `System.Web` dependencies that will need to be ported. The ApiPort tool was introduced in the last chapter on [understanding and updating dependencies](/understand-update-dependencies.md).

After [installing and configuring the *ApiPort* tool](https://docs.microsoft.com/dotnet/standard/analyzers/portability-analyzer), run the analysis from within Visual Studio, as shown in Figure 4-2.

![Figure 4-2](media/Figure4-2.png)

> **Figure 4-2. Analyze assembly portability in Visual Studio.**

Reviewer: ```suggestion
**Figure 4-2.** Analyze assembly portability in Visual Studio.
```

Choose the web project's assembly from the project's *bin* folder, as shown in Figure 4-3.

![Figure 4-3](media/Figure4-3.png)

> **Figure 4-3. Choose the project's web assembly.**

Reviewer: ```suggestion
**Figure 4-3.** Choose the project's web assembly.
```

![Figure 4-5](media/Figure4-5.png)

**Figure 4-5. *ApiPort* incompatible type details.**

Most of the incompatible types refer to `Controller` and various related attributes that have equivalents in ASP.NET Core.

## Update project files and NuGet reference syntax

Next, migrate from the older *.csproj* file structure to the newer, simpler structure introduced with .NET Core. In doing so, you'll also migrate from using a *packages.config* file for NuGet references to using `<PackageReference>` elements in the project file.

The original project's *eShopLegacyMVC.csproj* file is 418 lines long. A sample of the project file is shown in Figure 4-6. To offer a sense of its overall size and complexity, the right side of the image contains a miniature view of the entire file.

![Figure 4-6](media/Figure4-6.png)

> **Figure 4-6. The *eShopLegacyMVC.csproj* file structure.**

Reviewer: ```suggestion
**Figure 4-6.** The *eShopLegacyMVC.csproj* file structure.
```

In addition to the C# project file, NuGet dependencies are stored in a separate 45-line *packages.config* file, as shown in Figure 4-7.

![Figure 4-7](media/Figure4-7.png)

**Figure 4-7. The *packages.config* file.**

After upgrading to the new *.csproj* file format, you can migrate *packages.config* in class library projects using Visual Studio. This functionality doesn't work with ASP.NET projects, however. [Learn more about migrating *packages.config* to `<PackageReference>` in Visual Studio](https://docs.microsoft.com/nuget/consume-packages/migrate-packages-config-to-package-reference). If you have a large number of projects to migrate, [this community tool may help](https://github.com/MarkKharitonov/NuGetPCToPRMigrator).

## Create new ASP.NET Core project

Add a new ASP.NET Core project to the existing app's solution to make moving files easier, as most of the work can be done from within Visual Studio's **Solution Explorer**. In Visual Studio, right-click on your app's solution and choose **Add New Project**. Choose **ASP.NET Core web application**, and give the new project a name as shown in Figure 4-8.

![Figure 4-8](media/Figure4-8.png)

> **Figure 4-8. Add new ASP.NET Core web application.**

Reviewer: ```suggestion
**Figure 4-8.** Add new ASP.NET Core web application.
```

The next dialog will ask you to choose which template to use. Select the **Empty** template. Be sure to also change the dropdown from **.NET Core** to **.NET Framework**. Select **ASP.NET Core 2.2**, as shown in Figure 4-9.

![Figure 4-9](media/Figure4-9.png)

**Figure 4-9. Choose an Empty project template targeting .NET Framework with ASP.NET Core 2.2.**

### Migrating NuGet Packages

> Since the built-in migration tool for migrating *packages.config* to `<PackageReference>` doesn't work on ASP.NET projects, you can use a community tool instead, or migrate by hand. A community tool I've used is available here and uses an XSL file to transform from one format to the other. To use it, first copy the *packages.config* file to the newly created ASP.NET Core project folder. Make a backup of your files, as this script removes the *packages.config* file from all folders under where you run the script. Then run these commands from the project folder (or for the entire solution if you prefer):

Author: Fixed.

* *Content*
* *fonts*
* *Images*
* *Pics*
* *Scripts*

The **Empty** project template used in the previous step doesn't include this folder by default, or the middleware needed for it to work. You'll need to add them.

Add a *wwwroot* folder to the root of the project.

Add version 2.2.0 of the `Microsoft.AspNetCore.StaticFiles` NuGet package.

In *Startup.cs*, add a call to `app.UseStaticFiles()` in the `Configure` method:

```csharp
public void Configure(IApplicationBuilder app, IHostingEnvironment env)
{
    if (env.IsDevelopment())
    {
        app.UseDeveloperExceptionPage();
    }

    app.UseStaticFiles();
```

Copy the *Content* folder from the ASP.NET MVC app to the new project's *wwwroot* folder.

Run the app and navigate to its */Content/base.css* folder to verify that the static file is served correctly from its expected path. Continue copying the rest of the folders containing static files to the new project. You'll also want to copy the *favicon.ico* file from the project's root to the *wwwroot* folder. Figure 11 shows the results after these files and their folders have all been copied.

![Figure 4-11](media/Figure4-11.png)

> **Figure 4-11. Static folders copied over to *wwwroot* folder.**

Reviewer: ```suggestion
**Figure 4-11.** Static folders copied over to *wwwroot* folder.
```

First, modify the ported app's `appsettings.json` file and add the two settings in the root:

```json
{
  "Logging": {
    "LogLevel": {
      "Default": "Warning"
    }
  },
  "AllowedHosts": "*",
  "UseMockData": "true",
  "UseCustomizationData" :  "true"
}
```

Now, modify `ConfigureServices` to access the `UseMockData` setting from the `Configuration` property (where previously we set the value to `true`):

```csharp
  bool useMockData = Configuration.GetValue<bool>("UseMockData");
```

At this point, the setting is pulled from configuration. The other setting, `UseCustomizationData`, is used by the `CatalogDBInitializer` class. When you first ported this class, you commented out the access to `ConfigurationManager.AppSettings["UseCustomizationData"]`. Now it's time to modify it to use ASP.NET Core configuration. Modify the constructor of `CatalogDBInitializer` as follows:

```csharp
  // add using Microsoft.Extensions.Configuration
  public CatalogDBInitializer(CatalogItemHiLoGenerator indexGenerator,
      IConfiguration configuration)
  {

>       this.indexGenerator = indexGenerator;

Reviewer: I'd really like to clean this up, private fields should be prefixed with `_` to avoid using `this` as a qualifier.

Author: For now it's literally the code from the repo which uses the `this.` convention. I agree with you and prefer `_` but I don't necessarily want to have to reformat the whole sample project. And I think ideally we want these code listings to accurately reflect the referenced sample. What do you suggest?

Reviewer: I thought I saw the correct naming convention in other examples in this eBook, no? I could be mistaken. I'd really like to get this fixed, let's make it a post-publish effort.

The app's migration is nearly complete. The only remaining task is data access configuration.
  
## Data access considerations

ASP.NET Core apps running on .NET Framework can continue to leverage Entity Framework (EF). If performing an incremental migration, getting the app working with EF 6 before trying to port its data access to use EF Core may be worthwhile. In this way, any problems with the app's migration can be identified and addressed before another block of migration effort is begun.

As it happens, configuring EF 6 in the eShop sample migration doesn't require any special work, since this work was performed in the Autofac `ApplicationModule`. The only problem is that currently the `CatalogDBContext` class tries to read its connection string from *web.config*. To address this, the connection details need to be added to *appsettings.json*. Then the connection string must be passed into `CatalogDBContext` when it's created.

> First, *appsettings.json* is updated to include the connection string. The full file is listed here:

Reviewer: ```suggestion
Update the *appsettings.json* to include the connection string. The full file is listed here:
```

```json
{
  "ConnectionStrings": {
    "DefaultConnection": "Server=(localdb)\\mssqllocaldb;Database=eShopPorted;Trusted_Connection=True;MultipleActiveResultSets=true"
  },
  "Logging": {
    "LogLevel": {
      "Default": "Warning"
    }
  },
  "AllowedHosts": "*",
  "UseMockData": "false",
  "UseCustomizationData": "true"
}
```

> Next, the connection string must be passed into the constructor when the `DbContext` is created. Since the instances are created by Autofac, the change needs to be made in `ApplicationModule`. Modify the module to take in a `connectionString` in its constructor and assign it to a field. Then modify the registration for `CatalogDBContext` to add connection string as a parameter:

Reviewer: ```suggestion
The connection string must be passed into the constructor when the `DbContext` is created. Since the instances are created by Autofac, the change needs to be made in `ApplicationModule`. Modify the module to take in a `connectionString` in its constructor and assign it to a field. Then modify the registration for `CatalogDBContext` to add connection string as a parameter:
```

```csharp
protected override void OnModelCreating(ModelBuilder builder)
{
    builder.ApplyConfigurationsFromAssembly(Assembly.GetExecutingAssembly());

    base.OnModelCreating(builder);
}
```

Other changes involved include:

- `HasDatabaseGeneratedOption(DatabaseGeneratedOption.None)` replaced with `ValueGeneratedNever()`
- `HasRequired<T>` replaced with `HasOne<T>`
- Installed `Microsoft.EntityFrameworkCore.Relational` package
- Add a constructor to `CatalogDBContext` taking `DbContextOptions` and passing it to the base constructor

An example configuration class for `CatalogType` is shown here:

```csharp
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Metadata.Builders;

namespace eShopPorted.Models.Config
{
    public class CatalogTypeConfig : IEntityTypeConfiguration<CatalogType>
    {
        public void Configure(EntityTypeBuilder<CatalogType> builder)
        {

>             builder.ToTable("CatalogType");

Reviewer: ```suggestion
            builder.ToTable(nameof(CatalogType));
```

```csharp
dbContext.Database.EnsureDeleted();
dbContext.Database.EnsureCreated();
```

Seeding data in EF Core can be done with manual scripts, or as part of the type configuration. Along with other entity properties, seed data can be configured in `IEntityTypeConfiguration` classes by using `builder.HasData()`. The original app loaded seed data from CSV files in the *Setup* directory. Given that there are only a handful of items, these data records can instead be added as part of the entity configuration. This approach works well for lookup data in tables that change infrequently. Adding the following to `CatalogTypeConfig`'s `Configure` method ensures the associated rows are present when the database is created:

```csharp
builder.HasData(
    new CatalogType { Id = 1, Type = "Mug" },
    new CatalogType { Id = 2, Type = "T-Shirt" },
    new CatalogType { Id = 3, Type = "Sheet" },
    new CatalogType { Id = 4, Type = "USB Memory Stick" }
);
```

> The initial app includes a `PreconfiguredData` class, though, which includes data for `CatalogBrand` and `CatalogType`, so using this method the `HasData` call reduces to:

Reviewer: ```suggestion
The initial app includes a `PreconfiguredData` class, which includes data for `CatalogBrand` and `CatalogType`, so using this method the `HasData` call reduces to:
```

```csharp
builder.Property(ci => ci.Id)
    .ValueGeneratedNever()
    .IsRequired();
```

With these modifications, the ASP.NET Core app builds, but it doesn't yet work with EF Core, which must still be configured for dependency injection. With EF Core, the simplest way to configure it is in `ConfigureServices`:

```csharp
public IServiceProvider ConfigureServices(IServiceCollection services)
{
    services.AddMvc();
    bool useMockData = Configuration.GetValue<bool>("UseMockData");
    if (!useMockData)
    {
        string connectionString = Configuration.GetConnectionString("DefaultConnection");

        services.AddDbContext<CatalogDBContext>(options =>
            options.UseSqlServer(connectionString)
        );
    }

    // Create Autofac container builder
    var builder = new ContainerBuilder();
    builder.Populate(services);
    builder.RegisterModule(new ApplicationModule(useMockData));

    ILifetimeScope container = builder.Build();

    return new AutofacServiceProvider(container);
}
```

The final version of Autofac's `ApplicationModule` only configures one type, depending on whether the app is configured to use mock data:

```csharp

> public class ApplicationModule : Module
> {
>     private bool _useMockData;
> 
>     public ApplicationModule(bool useMockData)
>     {
>         _useMockData = useMockData;
>     }
>     protected override void Load(ContainerBuilder builder)
>     {
>         if (_useMockData)
>         {
>             builder.RegisterType<CatalogServiceMock>()
>                 .As<ICatalogService>()
>                 .SingleInstance();
>         }
>         else
>         {
>             builder.RegisterType<CatalogService>()
>                 .As<ICatalogService>()
>                 .InstancePerLifetimeScope();
>         }
>     }
> }

Reviewer: ```suggestion
public class ApplicationModule : Module
{
    private bool _useMockData;

    public ApplicationModule(bool useMockData)
    {
        _useMockData = useMockData;
    }
    
    protected override void Load(ContainerBuilder builder)
    {
        if (_useMockData)
        {
            builder.RegisterType<CatalogServiceMock>()
                .As<ICatalogService>()
                .SingleInstance();
        }
        else
        {
            builder.RegisterType<CatalogService>()
                .As<ICatalogService>()
                .InstancePerLifetimeScope();
        }
    }
}
```

```csharp
private void AddUriPlaceHolder(CatalogItem item)
{
    //TODO: Reference pic from wwwroot
    //item.PictureUri = this.Url.RouteUrl(PicController.GetPicRouteName, new { catalogItemId = item.Id }, this.Request.Url.Scheme);
}
```

The simplest fix is to reference the public image files in the site's public *wwwroot/Pics* directory. This task can be accomplished by replacing the method with the following code:

```csharp
private void AddUriPlaceHolder(CatalogItem item)
{
    item.PictureUri = $"/Pics/{item.Id}.png";
}
```

With this change, running the app reveals the images work as before.

## Additional MVC customizations

The *eShopLegacyMVC* app is fairly simple, so there isn't much to configure in terms of default MVC behavior. However, if you do need to configure additional MVC components, such as CORS, filters, and route constraints, you generally provide this information in `Startup.ConfigureServices`, where `UseMvc` is called. For example, the following code listing configures [CORS](https://docs.microsoft.com/aspnet/core/security/cors?view=aspnetcore-2.2&preserve-view=true) and sets up a global action filter:

```csharp

> public void ConfigureServices(IServiceCollection services)
> {
>     services.AddCors(options =>
>     {
>         options.AddPolicy(MyAllowSpecificOrigins,
>         builder =>
>         {
>             builder.WithOrigins("http://example.com",
>                                 "http://www.contoso.com")
>                                 .AllowAnyHeader()
>                                 .AllowAnyMethod();
>         });
>     });
> 
>     services.AddMvc(options =>
>     {
>       options.Filters.Add(new SampleGlobalActionFilter());
>     }).SetCompatibilityVersion(CompatibilityVersion.Version_2_2);
> }

Reviewer: ```suggestion
public void ConfigureServices(IServiceCollection services)
{
    services.AddCors(options =>
    {
        options.AddPolicy(MyAllowSpecificOrigins,
            builder =>
                builder.WithOrigins("http://example.com", "http://www.contoso.com")
                    .AllowAnyHeader()
                    .AllowAnyMethod());
    });

    services.AddMvc(options =>
    {
      options.Filters.Add(new SampleGlobalActionFilter());
    }).SetCompatibilityVersion(CompatibilityVersion.Version_2_2);
}
```

```

> **Note:** To finish configuring CORS, you must also call `app.UseCors()` in `Configure`.

Reviewer: ```suggestion
> [!Note]
> To finish configuring CORS, you must also call `app.UseCors()` in `Configure`.
```

Other advanced scenarios, like adding [custom model binders](https://docs.microsoft.com/aspnet/core/mvc/advanced/custom-model-binding?view=aspnetcore-2.2&preserve-view=true), formatters, and more are covered in the detailed ASP.NET Core docs. Generally these can be applied on an individual controller or action basis, or globally using the same options approach shown in the previous code listing.

## Other dependencies

Dependencies that use .NET Framework features that had a dependency on the legacy configuration model, such as the WCF client type and tracing code, must be modified when ported. Rather than having these types pull in their configuration information directly, they should be configured in code. For example, a connection to a WCF service that was configured in an ASP.NET app's *web.config* to use `basicHttpBinding` could instead be configured programmatically with the following code:

```csharp

> BasicHttpBinding binding = new BasicHttpBinding();

Reviewer: ```suggestion
var binding = new BasicHttpBinding();
```

> binding.MaxReceivedMessageSize = 2000000;

Reviewer: ```suggestion
binding.MaxReceivedMessageSize = 2_000_000;
```
